common/scala/src/main/scala/org/apache/openwhisk/core/database/elasticsearch/ElasticSearchActivationStore.scala

```diff
@@ -0,0 +1,467 @@
+  private val batcher: Batcher[IndexRequest, Either[ArtifactStoreException, DocInfo]] =
+    new Batcher(500, maxOpenDbRequests)(doStore(_)(TransactionId.dbBatcher))
+
+  private val minStart = 0L
+  private val maxStart = Instant.now.toEpochMilli + TimeUnit.DAYS.toMillis(365 * 100) //100 years from now
+
+  override def store(activation: WhiskActivation, context: UserContext)(
+    implicit transid: TransactionId,
+    notifier: Option[CacheChangeNotification]): Future[DocInfo] = {
+
+    val start =
+      transid.started(this, LoggingMarkers.DATABASE_SAVE, s"[PUT] 'activations' document: '${activation.docid}'")
+
+    val path = activation.annotations
+      .getAs[String](WhiskActivation.pathAnnotation)
+      .getOrElse(s"${activation.namespace}/${activation.name}")
+    // Escape `_id` field as it's not permitted in ElasticSearch, add `path` field for search, and
+    // convert annotations to JsObject as ElasticSearch doesn't support array with mixed types
+    // response.result can be any type ElasticSearch also doesn't support that, so convert it to a string
+    val response = JsObject(
+      activation.response.toJsonObject.fields
+        .updated("result", JsString(activation.response.result.toJson.compactPrint)))
+    val payload = JsObject(
+      activation.toDocumentRecord.fields - "_id" ++ Map(
+        "path" -> JsString(path),
+        "@timestamp" -> JsString(activation.start.toString),
+        "annotations" -> activation.annotations.toJsObject,
+        "response" -> response))
+
+    val index = generateIndex(activation.namespace.namespace)
+    val op = indexInto(index, esType).doc(payload.toString).id(activation.docid.asString)
+
+    val res = if (useBatching) {
+      batcher.put(op).map {
+        case Right(docInfo) =>
+          transid.finished(
+            this,
+            start,
+            s"[PUT] 'activations' completed document: '${activation.docid}', response: '$docInfo'")
+          docInfo
+        case Left(e: ArtifactStoreException) =>
+          transid.failed(
+            this,
+            start,
+            s"[PUT] 'activations' failed to put document: '${activation.docid}'; ${e.getMessage}.",
+            ErrorLevel)
+          throw PutException("error on 'put'")
+      }
+    } else {
+      client.execute(op).map { res =>
+        if (res.status == StatusCodes.OK.intValue || res.status == StatusCodes.Created.intValue) {
+          transid.finished(
+            this,
+            start,
+            s"[PUT] 'activations' completed activation: '${res.result.id}', response: '${activation.docinfo}'")
+          DocInfo(res.result.id)
+        } else {
+          transid.failed(
+            this,
+            start,
+            s"[PUT] 'activations' failed to put document: '${activation.docid}'; http status: '${res.status}'",
+            ErrorLevel)
+          throw new Exception("Unexpected http response code: " + res.status)
```

Comment:
Better to use a specific exception here. In general when store impls throw exception they throw exception extending `ArtifactStoreException` such that in `StoreUtils.reportFailure` those exception are not rehandled. In current impl `transid.failed` would be invoked twice and thus would result in wrong metrics being recorded

Reply:
oh, great point, sorry I missed it

about the odd coverage report, currently it only apply unittests from `org.apache.openwhisk.core.database.test.behavior.ActivationStoreBehavior`, in which `annotations` and `response.result` are all null, so `Try` will failed and fallbacks to `else` part, I will add some fake `result` and `annotations` to mock activations